storage/remote/read.go

```diff
@@ -278,3 +294,40 @@ func (sf seriesFilter) Labels() labels.Labels {
 	}
 	return labels
 }
+
+type lazySeriesSet struct {
```

Comment:
Don't we already have one of these?

Reply:
What we have implements `genericSeriesSet`, this one implements `storage.SeriesSet`.

Reply:
Why do we need two mildly different variants? I thought the whole purpose of all the generic/adapter complexity was to avoid duplication.

Reply:
I don't know the exact design decisions behind generic/adapter approach. 
But sure, same the behaviour could be achieved by exposing generic and adapter types and use it in here.
```go
&storage.SeriesSetAdapter{
		&storage.LazySeriesSet{create: func() (storage.GenericSeriesSet, bool) {
			set, ok := <-promise
			if !ok {
				return storage.SeriesSetAdapter{storage.ErrSeriesSet(errors.New("channel closed before a value received"))}, false
			}
			s := storage.SeriesSetAdapter{set}
			return s, s.Next()
		}}
```
However I don't think this is worth exposing all those types.

Reply:
@bwplotka How is this meant to work?

Reply:
Another option to keep a single `LazySeriesSet` and use `genericSeriesSetAdapter` to wrap it and use it in fanout.

Reply:
ok I had another look and don't think we can avoid a bit of duplication, because of the `init` function signature, but don't need the Err and Warning methods to can be like:

```
func LazySeriesSet(init func() (SeriesSet, bool)) SeriesSet {
	return &lazySeriesSet{init: init}
}

// lazySeriesSet is a wrapped series set that is initialised on first call to Next().
type lazySeriesSet struct {
	lazyGenericSeriesSet
	init func() (s SeriesSet, ok bool)

	set SeriesSet
}

func (c *lazySeriesSet) Next() bool {
	if c.set != nil {
		return c.set.Next()
	}

	var ok bool
	c.set, ok = c.init()
	return ok
}

func (c *lazySeriesSet) At() Series {
	if c.set != nil {
		return c.set.At()
	}
	return nil
}
```